### PR TITLE
chore: unit tests

### PR DESCRIPTION
**Pull request by peternhale**

### CODE REVIEW

File: src/tar.ts [hunk 1]
```diff
@@ -5,78 +5,80 @@ import {touch} from './util'
...
-  try {
-    await new Promise((resolve, reject) => {
-      const zlib = require('zlib')
-      const tar = require('tar-fs')
-      const crypto = require('crypto')
-      let shaValidated = false
-      let extracted = false
-      const check = () => shaValidated && extracted && resolve()
+export class Tar {
+  static async extract(stream: NodeJS.ReadableStream, basename: string, output: string, sha?: string) {
```
Comment (RodEsp) [inline src/tar.ts:9]: What's the point of making this a class?
Functions are simpler and easier to unit test than classes.

Comment (RodEsp) [reply src/tar.ts:9]: The function can be stubbed by itself without needing to make it a method in a class though. Unnecessary classes just muddle things up.

I whipped up a bare-bones example of stubbing a function that should be very similar to what you would need to do here at https://github.com/RodEsp/sinon-function-stub-example. Check it out and lemme know if that makes sense.

**Review by RodEsp — CHANGES_REQUESTED** [review]
See comment above about unnecessary classes.

File: src/tar.ts (hunk 1 above)
Comment (peternhale) [reply src/tar.ts:9]: @RodEsp I like your example. It seems to work with default export. How would this work when the exported function is not the default?

Comment (RodEsp) [reply src/tar.ts:9]: You'd need to stub the named export then, not the default.

Comment (peternhale) [reply src/tar.ts]: @RodEsp thanks, I have made the requested changes

**Comment by RodEsp** [comment]
What are the benefits of that pulling out most of the functionality in `src/commands/update.ts` and putting it in a new class in `src/update.ts`?

If the logic needs to be removed from `UpdateCommand` or refactored for unit testing that's perfectly fine, but I don't think it needs to be re-implemented in a separate class.

Functionality wise, this looks good but I'm concerned about the new `UpdateCli` class for a few reasons:
* The new class doesn't seem necessary for unit testing.
* The new class is tightly coupled with the `UpdateCommand` class because `UpdateCommand` just forwards its args and flags objects to the new class and the new class binds those to `this.options`. So if, for example, you want to see how/where any 2nd level `this.` property is defined in the new class you have to go look at it in `UpdateCommand`.
* The new class extends a new abstraction, the `EventEmitter` class, and I'm not sure what the benefit of that is either. Introducing a new abstraction like this is not necessary for unit testing which is the goal of this work.
* I did a diff on the main part of `UpdateCli` vs the old `UpdateCommand` and other than the removal of the legacy update logic nothing else seems to have really changed further making me think a new class is not necessary.
  * https://www.diffchecker.com/FSXmnmOG

**Comment by peternhale** [comment]
> What are the benefits of that pulling out most of the functionality in `src/commands/update.ts` and putting it in a new class in `src/update.ts`?
> 
> If the logic needs to be removed from `UpdateCommand` or refactored for unit testing that's perfectly fine, but I don't think it needs to be re-implemented in a separate class.
> 
> Functionality wise, this looks good but I'm concerned about the new `UpdateCli` class for a few reasons:
> 
> * The new class doesn't seem necessary for unit testing.
> * The new class is tightly coupled with the `UpdateCommand` class because `UpdateCommand` just forwards its args and flags objects to the new class and the new class binds those to `this.options`. So if, for example, you want to see how/where any 2nd level `this.` property is defined in the new class you have to go look at it in `UpdateCommand`.
> * The new class extends a new abstraction, the `EventEmitter` class, and I'm not sure what the benefit of that is either. Introducing a new abstraction like this is not necessary for unit testing which is the goal of this work.
> * I did a diff on the main part of `UpdateCli` vs the old `UpdateCommand` and other than the removal of the legacy update logic nothing else seems to have really changed further making me think a new class is not necessary.
>   
>   * https://www.diffchecker.com/FSXmnmOG

While I was out last week I had a feeling I would meet some questions regarding this PR ;).

This refactor was wholly done for unit testing. I was having trouble getting oclif command innards out of the way, so the solution in my mind was to move the logic to a class I had full control over.

Extending EventEmitter was done to avoid having to pass a reference of UX or to use `cli` directly. This is a similar challenge sfdx is tackling with providing user feedback directly from API like classes w/o having to use cli output directly. Using events seems like a reasonable choice.

I personally like writing very light weight command impls placing the bulk of the command's logic into other classes. I would like to continue doing so.

As for tight coupling, that is easily solved by expanding the set of available options to include specific properties for the needed data in the UpdateCli class.

I did have a conversation with Steve about unit tests for sfdx commands. The sfdx team has a number of test utilities that assist in unit testing commands, such as `@salesforce/ts-sinon`. If we are going to be challenged with making the current oclif unit tests real unit tests, it might be worthwhile discussing how to make some of these existing testing utils available to oclif unit tests.

**Comment by peternhale** [comment]
> I agree that light weight command implementations are best and ideally most of the logic they require will be placed in helper or utility files. I just don't think it needs to be placed in a class that introduces a new abstraction into this code base if the purpose if just to facilitate unit testing. It makes the code harder to understand and follow and I'd like to keep the code as readable, simple, and easy-to-understand as possible in Oclif repositories for the benefit of the wider OSS community.
> 
> In terms of "getting oclif command innards out of the way" have you looked at the testing utilities Oclif already has like https://github.com/oclif/test? Or looking at other Oclif plugins to see what they're doing? The [getCommands() function in plugin-commands](https://github.com/oclif/plugin-commands/blob/master/src/commands/commands.ts#L83) is a good example of a pattern that doesn't require extra classes but still allows you to "mock" oclif innards by stubbing out that method as shown in [its tests](https://github.com/oclif/plugin-commands/blob/master/test/commands/commands.test.ts#L65).

Why are you opposed to having the impl expressed as a class? Other option would be a number of global functions, which I am not fond of. Expressing the logic in a class or a set of global function in a helper file seems equivalent when applying the "just for unit testing" reason for doing this work.

The oclif/test framework, while simple to use, makes it difficult to mock functions in the command class given the way the test framework is running the target command. The fancy test route is what I am trying to replace with mocha, so that makes me feel like I am going backwards.

**Comment by RodEsp** [comment]
> Why are you opposed to having the impl expressed as a class?

Because classes in JavaScript are syntactical sugar that should be used when they help you reason about a program but not otherwise. Pulling functionality out of an existing class only to paste it into a new class that is tightly coupled to the original one doesn't do that. It will just create bloat that will be harder to work with and reason about in the future.

Similarly, EventEmitters are useful and should be used when there is a one to many relationship between the emitter and the subscribers but in this implementation that relationship is 1 to 1. Using this pattern is unnecessary and adding unnecessary abstractions to code makes it harder to work with and reason about as well.

You don't have to create global functions to pull out functionality into a different file or make it easier to unit test it. You can export functions as part of a module without it needing to be a class as shown in my [previous example](https://github.com/oclif/plugin-update/pull/325#discussion_r745178818).

Can you elaborate on "The oclif/test framework, while simple to use, makes it difficult to mock functions in the command class"?
[The example I linked to in plugin-command](https://github.com/oclif/plugin-commands/blob/master/test/commands/commands.test.ts#L65) shows how you can do this in a single line.

**Comment by peternhale** [comment]
> > Why are you opposed to having the impl expressed as a class?
> 
> Because classes in JavaScript are syntactical sugar that should be used when they help you reason about a program but not otherwise. Pulling functionality out of an existing class only to paste it into a new class that is tightly coupled to the original one doesn't do that. It will just create bloat that will be harder to work with and reason about in the future.
> 
> Similarly, EventEmitters are useful and should be used when there is a one to many relationship between the emitter and the subscribers but in this implementation that relationship is 1 to 1. Using this pattern is unnecessary and adding unnecessary abstractions to code makes it harder to work with and reason about as well.
> 
> You don't have to create global functions to pull out functionality into a different file or make it easier to unit test it. You can export functions as part of a module without it needing to be a class as shown in my [previous example](https://github.com/oclif/plugin-update/pull/325#discussion_r745178818).
> 
> Can you elaborate on "The oclif/test framework, while simple to use, makes it difficult to mock functions in the command class"? [The example I linked to in plugin-command](https://github.com/oclif/plugin-commands/blob/master/test/commands/commands.test.ts#L65) shows how you can do this in a single line.

- `oclif/test framework` my first attempt at this was to use the fancy test tests that already existed and mock all of the http calls being made. Attempts to do this failed, as if the mocks were not being applied. When I debugged into the issue I found that the tests were being run in a separate process and the mocks I made were not being applied. Moving the logic to a class I can control, both in setup and execution, solves this problem. Not sure if I did something wrong, but Steve has seen similar issues with fancy test and mocks.
Fancy test works really when when it can stay within the node process, but when the command is doing lots of other stuff, it gets messy, IMHO.
- I misused the term global. I was referring to independent exported functions.
- Not sure how much bloat is added by exposing the logic in the command as a class.

**Pull request closed without merging**